core/src/main/java/org/elasticsearch/discovery/zen/NodeJoinController.java

```diff
@@ -376,20 +378,26 @@ public ClusterState execute(ClusterState currentState) {
                 while (iterator.hasNext()) {
                     Map.Entry<DiscoveryNode, List<MembershipAction.JoinCallback>> entry = iterator.next();
                     final DiscoveryNode node = entry.getKey();
-                    joinCallbacksToRespondTo.addAll(entry.getValue());
                     iterator.remove();
-                    if (currentState.nodes().nodeExists(node.getId())) {
+                    final String preflight = nodesBuilder.preflightPut(node);
+                    if (preflight != null) {
+                        Throwable failure = new IllegalStateException(preflight);
+                        for (MembershipAction.JoinCallback callback: entry.getValue()) {
+                            joinCallbacksToFail.add(new Tuple<>(callback, failure));
+                        }
+                    } else if (currentState.nodes().nodeExists(node.getId())) {
```

Comment:
That's part of the exception the joining nodes gets? It will be logged [here](https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/discovery/zen/ZenDiscovery.java#L488)